test-sbt/shared/src/main/scala/zio/test/runner/ZSpecStructure.scala

```diff
@@ -0,0 +1,87 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio.test.runner
+
+import zio.{ RIO, ZIO }
+import zio.test.Spec.{ SuiteCase, TestCase }
+import zio.test.runner.ExecutedSpecStructure.{ ExecutedSuite, ExecutedTest, Stats }
+import zio.test.{ Assertion, ExecutedSpec, FailureDetails, TestResult }
+
+sealed abstract class ExecutedSpecStructure {
+  def stats: Stats
+  def traverse[R](
+    handleSuite: (String, Stats) => RIO[R, Unit],
+    handleTestResult: (String, ZTestResult) => RIO[R, Unit]
+  ): RIO[R, Unit] =
+    this match {
+      case ExecutedSuite(label, stats, children) =>
+        for {
+          _ <- ZIO.sequence(children.map(_.traverse(handleSuite, handleTestResult)))
+          _ <- handleSuite(label, stats)
+        } yield ()
+      case ExecutedTest(label, result) => handleTestResult(label, result)
+    }
+}
+
+object ExecutedSpecStructure {
+  case class Stats(passed: Int = 0, failed: Int = 0, ignored: Int = 0) {
+    val total: Int = passed + failed + ignored
+    def +(that: Stats): Stats = copy(
+      passed = passed + that.passed,
+      failed = failed + that.failed,
+      ignored = ignored + that.ignored
+    )
+  }
+
+  def from[L, R](spec: ExecutedSpec[L]): ExecutedSpecStructure =
+    spec.caseValue match {
+      case SuiteCase(label, specs, _) =>
+        val children = specs.map(from)
+        val stats    = children.map(_.stats).foldLeft(Stats())(_ + _)
+        ExecutedSuite(label.toString, stats, children)
+
+      case TestCase(label, testResult) =>
+        ExecutedTest(label.toString, ZTestResult.from(testResult))
+    }
+
+  case class ExecutedSuite(label: String, stats: Stats, children: Vector[ExecutedSpecStructure])
+      extends ExecutedSpecStructure
+  case class ExecutedTest(label: String, result: ZTestResult) extends ExecutedSpecStructure {
+    override def stats: Stats = result.stats
+  }
+}
+
+sealed abstract class ZTestResult(val stats: Stats) {
+  def rendered = toString.toUpperCase
+}
+object ZTestResult {
+  case object Ignored                                                        extends ZTestResult(Stats(ignored = 1))
+  case object Success                                                        extends ZTestResult(Stats(passed = 1))
+  case class Failure(details: FailureDetails, override val rendered: String) extends ZTestResult(Stats(failed = 1))
+
+  def from(testResult: TestResult): ZTestResult =
+    testResult match {
+      case Assertion.Success     => Success
+      case Assertion.Failure(fd) => Failure(fd, describe(fd))
+      case Assertion.Ignore      => Ignored
+    }
+
+  private def describe(failureDetails: FailureDetails) = failureDetails match {
```

Comment:
This message will be the least helpful one (`whole`). The `fragment` one will be more relevant. 

However, it would be great if we could use the existing `TestReporter` to render this. Maybe if we rendered into an intermediate form first, then you could grab that and use it here? What do you think?

Reply:
Yes, it would be great to reuse the TestReporter.
When I wrote this the reporter didn't exist yet, but when I saw it a copied the `did not satisfy phrasing` :)

Reply:
Exposed rendering to a  intermediate form in TestReporter and reused